src/pip/_internal/resolution/legacy/resolver.py

```diff
@@ -591,7 +592,7 @@ def schedule(req: InstallRequirement) -> None:
             if req.constraint:
                 return
             ordered_reqs.add(req)
-            for dep in self._discovered_dependencies[req.name]:
+            for dep in self._discovered_dependencies[req.name]:  # type: ignore[index]
```

Comment:
What’s the case this ignores?

Reply:
`self._discovered_dependencies` is a `DefaultDict[str, List[InstallRequirement]]`, but mypy believes `req.name` could be None. Like I mentioned in the body of the PR:

> I can't prove that req.name is not None in get_installation_order and the code would work fine if that were the case (since it's a defaultdict), so I opted to just ignore

Reply:
Yes, we can. But I wanted to make it clear in code review that I have not been able to prove that the assert is "correct" and that the code would "work" as is.

Was hoping someone could chime in — first and foremost, introduce no new bugs.

mypy thinks that because the type is added here:
https://github.com/pypa/pip/blob/324dd444956283661dce0dc282cbdaad0405d921/src/pip/_internal/resolution/legacy/resolver.py#L156
The `DiscoveredDependencies` type alias is defined here:
https://github.com/pypa/pip/blob/324dd444956283661dce0dc282cbdaad0405d921/src/pip/_internal/resolution/legacy/resolver.py#L54-L55

Reply:
Whoops. I missed that was a type alias. My bad. Maybe we should just change `str` in that to `Any` (or `Optional[str]` if that's the type of `req.name`) given that inserting non-str values isn't going to cause any problem?

Reply:
+1 to making the key type `Optional[str]`, I don’t think there’s any downsides.